File: backend/app/routers/clientes.py
"""
Endpoints FastAPI — Clientes
Equivalente a: FrmPrincipal.vb, frmClienteNovo.vb e frmAlterar.vb
(Solo Consultoria de Imóveis)
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.schemas.clientes import (
    ClienteCreate,
    ClienteCreateResponse,
    ClienteDetail,
    ClientesResponse,
    ClienteUpdate,
    ClienteUpdateResponse,
    ProximoCodigoResponse,
    VerificarSenhaRequest,
    VerificarSenhaResponse,
)
from app.services.clientes import (
    atualizar_cliente,
    buscar_cliente,
    criar_cliente,
    excluir_cliente,
    listar_clientes,
    proximo_codigo,
    verificar_senha_exclusao,
)

router = APIRouter(prefix="/clientes", tags=["Clientes"])


@router.get(
    "/",
    response_model=ClientesResponse,
    summary="Lista todos os clientes",
    description=(
        "Retorna a lista completa de clientes ordenada por nome. "
        "Equivalente ao ClientesTableAdapter.Fill() no "
        "FrmPrincipal_Load."
    ),
)
async def get_clientes(
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN09 — Grid de clientes carregado do banco, ordenado por nome.
    RN10 — O primeiro item da lista fornece o código inicial
           exibido no header da seção de ações.
    """
    try:
        clientes = listar_clientes(db)
        return ClientesResponse(clientes=clientes)
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erro ao carregar clientes: {str(e)}",
        )


# ---------------------------------------------------------------------------
# frmClienteNovo — endpoints de criação
# ---------------------------------------------------------------------------


@router.get(
    "/proximo-codigo",
    response_model=ProximoCodigoResponse,
    summary="Próximo código disponível",
    description=(
        "Retorna o primeiro código disponível >= 10.000 e o próximo Id "
        "para criação de novo cliente. Equivalente ao btnCodigo_Click no "
        "frmClienteNovo."
    ),
)
async def get_proximo_codigo(
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN23 — Geração automática: primeiro inteiro >= 10.000 não utilizado.
    RN27 — Retorna também o próximo Id disponível.
    """
    return proximo_codigo(db)


@router.post(
    "/",
    response_model=ClienteCreateResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Cria novo cliente",
    description=(
        "Cria um novo cliente com validações de código único, "
        "limite de 20.000 e campos obrigatórios. Equivalente ao "
        "btnRibSalvar_Click no frmClienteNovo."
    ),
)
async def post_cliente(
    body: ClienteCreate,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN21 — Código deve ser único na tabela Clientes.
    RN22 — Código máximo: 20.000.
    RN24 — Campo Cliente obrigatório.
    RN25 — Nome convertido para maiúsculas.
    RN26 — Campo Código obrigatório.
    RN29 — Código duplicado: retorna 409 "Código já existente".
    RN30 — Código > 20.000: retorna 422.
    """
    return criar_cliente(db, codigo=body.codigo, cliente=body.cliente)


# ---------------------------------------------------------------------------
# frmAlterar — endpoints de edição e exclusão
# ---------------------------------------------------------------------------


@router.get(
    "/{id}",
    response_model=ClienteDetail,
    summary="Busca cliente por Id",
    description=(
        "Retorna dados de um cliente específico e indica se possui "
        "lançamentos em Contas. Equivalente ao frmAlterar_Load "
        "(Fill + Filter por Id)."
    ),
)
async def get_cliente(
    id: int,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN33 — Carrega dados do cliente pelo Id.
    RN42 — Verifica existência de lançamentos (tem_lancamentos).
    """
    return buscar_cliente(db, id)


@router.put(
    "/{id}",
    response_model=ClienteUpdateResponse,
    summary="Atualiza cliente",
    description=(
        "Atualiza Código e/ou Nome do cliente. Validações: código "
        "único (excluindo o próprio Id), nome em maiúsculas. "
        "Equivalente a TableAdapterManager.UpdateAll() no "
        "btnRibSalvar_Click do frmAlterar."
    ),
)
async def put_cliente(
    id: int,
    body: ClienteUpdate,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN36 — Alteração de Código (confirmação de alto risco no frontend).
    RN37 — Alteração de Nome (confirmação simples no frontend).
    RN38 — Após salvar, frontend retorna campos a ReadOnly.
    RN47 — Nome convertido para maiúsculas.
    RN48 — Código deve ser único excluindo o próprio Id.
    """
    return atualizar_cliente(db, id=id, codigo=body.codigo, cliente=body.cliente)


@router.post(
    "/{id}/verificar-senha",
    response_model=VerificarSenhaResponse,
    summary="Verifica senha de exclusão",
    description=(
        "Valida a senha fornecida contra a tabela Chaves "
        "(Ref='Exclusão de cliente'). Retorna também se o cliente "
        "possui lançamentos, para o frontend decidir se exibe "
        "confirmação adicional. Equivalente a frmSenha (varSenha=2)."
    ),
)
async def post_verificar_senha(
    id: int,
    body: VerificarSenhaRequest,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN41 — Senha validada contra Chaves WHERE Ref='Exclusão de cliente'.
    RN42 — Retorna tem_lancamentos para controle de fluxo no frontend.
    """
    return verificar_senha_exclusao(db, id=id, senha=body.senha)


@router.delete(
    "/{id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Exclui cliente",
    description=(
        "Exclui o cliente e todos os seus lançamentos em Contas "
        "(se houver), em transação. Equivalente ao código ADODB "
        "comentado no btnExcluir_Click do frmAlterar."
    ),
)
async def delete_cliente(
    id: int,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """
    RN43 — Confirmação adicional (se tem lançamentos) validada no frontend.
    RN44 — DELETE Contas + DELETE Clientes em transação.
    RN45 — DELETE Clientes apenas se sem lançamentos.
    RN46 — Frontend invalida cache e navega para /principal após sucesso.
    """
    excluir_cliente(db, id=id)

File: frontend/src/services/clientes.ts
/**
 * services/clientes.ts
 * Camada de acesso à API de clientes.
 * Única camada que conhece as URLs e o contrato HTTP do backend.
 *
 * Usa apiFetch (autenticado) — diferente de auth.ts que usa fetch nativo.
 */

import { apiFetch } from "@/lib/api";

// ---------------------------------------------------------------------------
// Tipos — alinhados aos schemas Pydantic do backend (schemas/clientes.py)
// ---------------------------------------------------------------------------

export interface ClienteListItem {
  id: number;
  codigo: number;
  cliente: string;
}

interface ClientesResponse {
  clientes: ClienteListItem[];
}

export interface ProximoCodigoResponse {
  proximo_codigo: number;
  proximo_id: number;
}

export interface ClienteCreateRequest {
  codigo: number;
  cliente: string;
}

export interface ClienteCreateResponse {
  id: number;
  codigo: number;
  cliente: string;
}

// ---------------------------------------------------------------------------
// Funções de serviço
// ---------------------------------------------------------------------------

/**
 * Busca a lista completa de clientes, ordenada por nome.
 * Equivalente ao ClientesTableAdapter.Fill() no FrmPrincipal_Load.
 * RN09 — Lista carregada do banco, ordenada por nome.
 */
export async function fetchClientes(): Promise<ClienteListItem[]> {
  const res = await apiFetch("/clientes/");

  if (!res.ok) {
    const err = await res.json();
    throw new Error(
      err.detail ?? "Falha ao carregar clientes",
    );
  }

  const data: ClientesResponse = await res.json();
  return data.clientes;
}

/**
 * Gera o próximo código disponível >= 10.000.
 * Equivalente ao btnCodigo_Click no frmClienteNovo.
 * RN23 — Primeiro inteiro >= 10.000 não utilizado.
 * RN27 — Retorna também o próximo Id.
 */
export async function fetchProximoCodigo(): Promise<ProximoCodigoResponse> {
  const res = await apiFetch("/clientes/proximo-codigo");

  if (!res.ok) {
    const err = await res.json();
    throw new Error(
      err.detail ?? "Falha ao gerar próximo código",
    );
  }

  return res.json();
}

/**
 * Cria um novo cliente.
 * Equivalente ao btnRibSalvar_Click no frmClienteNovo.
 * RN21, RN22, RN24, RN25, RN26.
 */
export async function criarCliente(
  data: ClienteCreateRequest,
): Promise<ClienteCreateResponse> {
  const res = await apiFetch("/clientes/", {
    method: "POST",
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(
      err.detail ?? "Falha ao criar cliente",
    );
  }

  return res.json();
}

// ---------------------------------------------------------------------------
// frmAlterar — funções de edição e exclusão
// ---------------------------------------------------------------------------

export interface ClienteDetail {
  id: number;
  codigo: number;
  cliente: string;
  /** RN42 — indica se há lançamentos em Contas para o cliente */
  tem_lancamentos: boolean;
}

export interface ClienteUpdateRequest {
  codigo: number;
  cliente: string;
}

export interface ClienteUpdateResponse {
  id: number;
  codigo: number;
  cliente: string;
}

export interface VerificarSenhaResponse {
  /** RN41 — True se senha confere com Chaves WHERE Ref='Exclusão de cliente' */
  valido: boolean;
  /** RN42 — True se cliente tem registros em Contas */
  tem_lancamentos: boolean;
}

/**
 * Busca dados de um cliente pelo Id.
 * Equivalente ao frmAlterar_Load (Fill + Filter por Id).
 * RN33, RN42.
 */
export async function fetchCliente(id: number): Promise<ClienteDetail> {
  const res = await apiFetch(`/clientes/${id}`);

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail ?? "Falha ao carregar cliente");
  }

  return res.json();
}

/**
 * Atualiza Código e/ou Nome do cliente.
 * Equivalente a TableAdapterManager.UpdateAll() no btnRibSalvar_Click.
 * RN36, RN37, RN47, RN48.
 */
export async function atualizarCliente(
  id: number,
  data: ClienteUpdateRequest,
): Promise<ClienteUpdateResponse> {
  const res = await apiFetch(`/clientes/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail ?? "Falha ao atualizar cliente");
  }

  return res.json();
}

/**
 * Valida a senha de exclusão contra a tabela Chaves.
 * Equivalente a frmSenha.ShowDialog() com varSenha="2".
 * RN41, RN42.
 */
export async function verificarSenhaExclusao(
  id: number,
  senha: string,
): Promise<VerificarSenhaResponse> {
  const res = await apiFetch(`/clientes/${id}/verificar-senha`, {
    method: "POST",
    body: JSON.stringify({ senha }),
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail ?? "Falha ao verificar senha");
  }

  return res.json();
}

/**
 * Exclui o cliente e seus lançamentos (se houver).
 * Equivalente ao código ADODB comentado no btnExcluir_Click.
 * RN44, RN45.
 */
export async function excluirCliente(id: number): Promise<void> {
  const res = await apiFetch(`/clientes/${id}`, {
    method: "DELETE",
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail ?? "Falha ao excluir cliente");
  }
}
